jsonschema/src/compilation/mod.rs

```diff
@@ -588,51 +583,63 @@ mod tests {
         // Define a custom keyword validator that wraps "minimum"
         // but maintains a counter of how many times the validator was applied.
         struct CountingValidator {
-            count: u32,
+            count: Arc<Mutex<i64>>,
         }
-        impl<'instance> CustomKeywordValidator<'instance, '_> for CountingValidator {
-            fn validate(
+
+        impl CountingValidator {
+            fn increment(&mut self, amount: i64) {
+                let mut count = self.count.lock().expect("Lock is poisoned");
+                *count += amount;
+            }
+        }
+
+        impl CustomKeywordValidator for CountingValidator {
+            fn validate<'instance>(
                 &mut self,
-                instance: &'instance Value,
-                instance_path: JSONPointer,
+                _: &'instance Value,
+                _: JSONPointer,
                 subschema: Arc<Value>,
-                subschema_path: JSONPointer,
-                schema: Arc<Value>,
+                _: JSONPointer,
+                _: Arc<Value>,
             ) -> ErrorIterator<'instance> {
                 let amount = match &*subschema {
                     Value::Number(x) => x.as_i64().expect("countme value must be integer"),
                     _ => panic!("Validator requires numeric values"),
                 };
-                self.count += amount as u32;
+                self.increment(amount);
                 Box::new(None.into_iter())
             }
 
-            fn is_valid(&mut self, instance: &Value, subschema: &Value, schema: &Value) -> bool {
+            fn is_valid(&mut self, _: &Value, subschema: &Value, _: &Value) -> bool {
                 let amount = match subschema {
                     Value::Number(x) => x.as_i64().expect("countme value must be integer"),
                     _ => return false,
                 };
-                self.count += amount as u32;
+                self.increment(amount);
                 true
             }
         }
 
         // define compilation options that include the custom format and the overridden keyword
-        let validator = Arc::new(Mutex::new(Box::new(CountingValidator { count: 0 })));
+        let count = Arc::new(Mutex::new(0));
+        let boxed: Box<dyn CustomKeywordValidator> = Box::new(CountingValidator {
```

Comment:
It's interesting that Rust can't preserve that this is a `CountingValidator` while passing it to a function that expects a `dyn CustomKeywordValidator`. That's what I was getting hung up on. The inner state being behind another Arc<Mutex<T>> seems like a good alternative.

Reply:
Technically, Rust preserves this information, but behind an indirection. Dynamic dispatch "knows" at runtime what kind of object is there, so it can do a vtable lookup and find the relevant implementation of methods from `CustomKeywordValidator`. It is also possible to downcast `dyn CustomKeywordValidator` to `CountingValidator` via `downcast_ref`, though it is not that convenient to use.

Without this kind of indirection, it would not be clear how much space to allocate on the stack for that type + as we store it in a hashmap, we can only use boxed trait objects.